File: 0543-diameter-of-binary-tree/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int height(struct TreeNode* root){
    if(root==NULL)return 0;
    return 1+fmax(height(root->left),height(root->right));
}
int diameterOfBinaryTree(struct TreeNode* root) {
    if(root==NULL)return 0;
    int lh=height(root->left);
    int rh=height(root->right);
    int lw=diameterOfBinaryTree(root->left);
    int rw=diameterOfBinaryTree(root->right);
    return fmax(lh+rh,fmax(lw,rw));
}

File: 0543-diameter-of-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int ans =0;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int c = q.size();
            while(c>0){
                c--;
                TreeNode cur = q.poll();
                ans=Math.max(ans,height(cur.left)+height(cur.right));
                if(cur.left!=null)q.add(cur.left);
                if(cur.right!=null)q.add(cur.right);
            }
        }
        return ans;
    }
    public int height(TreeNode root){
        if(root==null){
            return 0;
        }
        return 1+Math.max(height(root.left),height(root.right));
    }

}
